rfcs/framework-0063-notification-system.md

```diff
@@ -0,0 +1,422 @@
+
+It will take care of sending the notifications based on the configuration and user preferences (i.e. if the repository disables certain notifications or a user does not want to receive certain notifications, they should not be sent).
+
+### Use Case
+
+- Send notifications for operations involving multiple users (see [table](#notification-examples))
+- Overridable notification templates
+  - One template for each notification
+  - Can be overridden via theme (like other templates)
+  - Add new templates for custom notifications
+  - Should be translatable
+    - Store preferred language in user profile
+- Modify notifications
+  - Set which notifications should be send
+    - As a repository admin activate/deactive certain notifications
+    - As a user (user profile)
+  - Define notifications for events per module
+  - Define notification backends per module
+
+## Detailed design
+
+In this design, we describe:
+
+- New `invenio-notifications` module
+    - Simple self-contained API that is business-logic agnostic
+    - Minimal dependencies, only for supporting different notification transfer backends (email, chat, etc.)
+    - Defines a common notification payload/datamodel
+- Utilities for configuration on services for building notifications payloads (receivers, data payload)
+    - Generates the list of recipients
+    - Takes into account user preferences, e.g.
+        - if they want to receive notifications at all
+        - what kind of notifications the want to receive
+
+### Datamodel/payload
+
+- Acts as a "client payload"
+- Keeps all the relevant data necessary for a notification to be dispatched via a backend
+- It must be JSON-serializeable so that it can be passed to Celery tasks.
+
+```python
+@dataclass
+class Notification:
+    """Notification datamodel."""
+
+    # Unique type identifier for the notification
+    type: str
+
+    # Actual subject/payload of the notification's objects, e.g. record,
+    # community, request, etc.
+    context: dict
+
+    # Recipients list
+    recipients: list[Recipient]
+
+    # Creation time of the notification
+    timestamp: datetime
+
+
+@dataclass
+class Recipient:
+    """Recipient datamodel."""
+
+    # Recipient information (e.g. expanded user/group)
+    data: dict
+
+    # The "rendering" template to be used for this recipient
+    template: str
+
+    # Backend delivery options/config
+    backends: list[Backend]
+
+
+@dataclass
+class Backend:
+    """Backend delivery datamodel."""
+
+    # Unique backend identifier ("email", "slack", "cern-notification")
+    id: str
+
+    # Options/parameters for the backend to deliver the notification
+    params: dict
+```
+
+### API
+
+A Notification manager/client will be used for sending notifications as a Celery task.
+
+```python
+class NotificationManager:
+
+    def __init__(self, backends):
+        self.backends = backends
+
+    def validate(self, notification):
```

Comment:
would validate raise exceptions or return `True` or `False`? this would affect the `broadcast` function below.

Reply:
This is meant to be called by the service and validates the produced notification payload, e.g. if required backend parameters are there (`email` always needs a `to` field, etc.).

The issue is that it's called in the `on_post_commit` clause of a UoW operation, so it should have the same behavior as e.g. failure of ElasticSearch indexing or sending a Celery task... In REST API this is a 500 response :)